File: 126.cpp
#include<iostream>
#include<queue>
#include<vector>
#include<string>
#include<unordered_set>

using namespace std;

class Solution {
public:
    vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
        vector<vector<string>> res;//all the paths
        unordered_set<string> dict(wordList.begin(),wordList.end());
        queue<vector<string>> paths;
        paths.push({beginWord});
        unordered_set<string> word;
        int level=1,minLevel=INT_MAX;

        while(!paths.empty()){
            auto cur=paths.front();
            paths.pop();
            while(cur.size()>level){
                for(auto&& w:word){
                    dict.erase(w);
                }
                level=cur.size();
                if(level>minLevel)//prune the search tree
                    break;
            }
            string last=cur.back();
            for(int i=0;i<last.size();++i){
                char x=last[i];
                for(char ch='a';ch<='z';++ch){
                    last[i]=ch;
                    if(dict.count(last)){
                        vector<string> next=cur;
                        word.insert(last);
                        next.push_back(last);
                        if(last==endWord){
                            res.push_back(next);
                            minLevel=level;
                        }
                        else
                            paths.push(next);
                    }
                }
                last[i]=x;
            }
        }
        return res;
    }
};

int main(){
}

File: Python/126.py
from collections import deque
class Solution(object):
    def findLadders(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: List[List[str]]
        """
        paths = deque()
        paths.append([beginWord])
        res = []
        words = set(wordList)
        level, minLevel = 1, len(words)
        levelWords = set()
        
        while paths:
            cur = paths.popleft()
            if len(cur) > level:
                level = len(cur)
                words = words - levelWords
                if level > minLevel:
                    break
            begin = cur[-1]
            for i in range(len(begin)):
                x = begin[i]
                for j in range(26):
                    ch = chr(ord('a')+j)
                    begin = begin[:i]+ch+begin[i+1:]
                    if begin in words:
                        levelWords.add(begin)
                        new_path = cur + [begin]
                        paths.append(new_path)
                        if begin == endWord:
                            minLevel = level
                            res.append(new_path)
                begin = begin[:i]+x+begin[i+1:]
            
        return res
